### PR TITLE
feat: Launcher argument for Fluent exposure and graphics driver

### PR DESCRIPTION
**Pull request by mkundu1**
Two new arguments `ui` and `graphics_driver` are added to `launch_fluent`. They accepts the values of `FluentUI` and `FluentWindowsGraphicsDriver | FluentLinuxGraphicsDriver` enums respectively. The default value of `ui` is `FluentUI.HIDDEN_GUI` (`-hidden`) in windows (no change) and `FluentUI.NO_GUI` (`-gu`) in linux. I haven't made the default value `FluentUI.NO_GUI` in windows as it shows Fluent output in the current cmd. The default value of `graphics_driver` is `AUTO`.

The `launch_fluent` argument `show_gui` has been made deprected.

TODO: Record `ui` and `graphics_driver` Python joural saved from Fluent.

### CODE REVIEW

File: src/ansys/fluent/core/launcher/launcher.py [hunk 1]
```diff
@@ -233,6 +257,27 @@ def launch_fluent(
...
+    del show_gui
+    if exposure is None:
+        # Not using NO_GUI in windows as it opens a new cmd or
+        # shows Fluent output in the current cmd if start is not used
+        exposure = FluentExposure.HIDDEN_GUI if _is_windows() else FluentExposure.NO_GUI
+    if isinstance(exposure, str):
+        exposure = FluentExposure(exposure)
+    if graphics_driver is None:
+        graphics_driver = "auto"
+    graphics_driver = str(graphics_driver)
```
Comment (mkundu1) [inline src/ansys/fluent/core/launcher/launcher.py:275]: Converting possible enum value to string to catch thigs like specifying `FluentLinuxGraphicsDriver` in windows during the enum constructor in the following line.

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 2]
```diff
@@ -179,6 +180,70 @@ def is_meshing(mode: "FluentMode") -> bool:
         return mode in [FluentMode.MESHING_MODE, FluentMode.PURE_MESHING_MODE]
 
 
+@total_ordering
+class FluentEnum(Enum):
+    """Base class for Fluent-related enums. Accepts lowercase member names as values
+    and supports comparison operators.
```
Comment (mkundu1) [inline src/ansys/fluent/core/launcher/launcher_utils.py:186]: test_fluent_enums for the supported behaviour

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 3]
```diff
@@ -282,30 +353,6 @@ def _get_mode(mode: Optional[Union[FluentMode, str, None]] = None):
     return mode
 
 
-def _raise_exception_g_gu_in_windows_os(additional_arguments: str) -> None:
```
Comment (mkundu1) [inline src/ansys/fluent/core/launcher/launcher_utils.py:285]: This is now supported in windows. (TODO Fluent version check)

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 4]
```diff
@@ -282,30 +353,6 @@ def _get_mode(mode: Optional[Union[FluentMode, str, None]] = None):
...
-    """If -g or -gu is passed in Windows OS, the exception should be raised."""
-    additional_arg_list = additional_arguments.split()
-    if _is_windows() and (
-        ("-g" in additional_arg_list) or ("-gu" in additional_arg_list)
-    ):
-        raise InvalidArgument("Unsupported '-g' and '-gu' on windows platform.")
-
-
-def _update_launch_string_wrt_gui_options(
```
Comment (mkundu1) [inline src/ansys/fluent/core/launcher/launcher_utils.py:294]: I have removed this function as user will now use the `exposure` argument instaed of `additional_arguments` for `-g` or `-gu`.

File: src/ansys/fluent/core/solver/flobject.py [hunk 5]
```diff
@@ -493,7 +493,6 @@ def _show_warning(message, category, *args, **kwargs):
 
 
 warnings.showwarning = _show_warning
-warnings.simplefilter("default", category=DeprecatedSettingWarning)
```
Comment (mkundu1) [inline src/ansys/fluent/core/solver/flobject.py:496]: `DeprecationWarning` from a module is ignored by default, but `FutureWarning` (recommeded as user-facing warning) is not.

File: src/ansys/fluent/core/warnings.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class PyFluentDeprecationWarning(FutureWarning):
```
Comment (mkundu1) [inline src/ansys/fluent/core/warnings.py:1]: All PyFluent deprecation warnings should use this class so that all such warnings can be disabled by the following code if user needs it: `warnings.filterwarnings("ignore", category=PyFluentDeprecationWarning)`

**Review by hpohekar — APPROVED** [review]

**Review by prmukherj — APPROVED** [review]

File: src/ansys/fluent/core/launcher/container_launcher.py [hunk 7]
```diff
@@ -59,6 +63,13 @@ def __init__(
...
+        exposure : FluentExposure
+            Sets the exposure of Fluent. The possible values are the values of the
+            ``FluentExposure`` enum.
+        graphics_driver : FluentWindowsGraphicsDriver or FluentLinuxGraphicsDriver
+            Sets the graphics driver of Fluent. The possible values are the values of
+            the ``FluentWindowsGraphicsDriver`` enum in Windows or the values of
+            the ``FluentLinuxGraphicsDriver`` enum in Linux.
         product_version : str, optional
             Select an installed version of ANSYS. The string must be in a format like
             ``"23.2.0"`` (for 2023 R2) matching the documented version format in the
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/container_launcher.py:75]:
```suggestion
            Exposure of Fluent. Options are the values of the
            ``FluentExposure`` enum.
        graphics_driver : FluentWindowsGraphicsDriver or FluentLinuxGraphicsDriver
            Graphics driver of Fluent. Options are the values of
            the ``FluentWindowsGraphicsDriver`` enum in Windows or the values of
            the ``FluentLinuxGraphicsDriver`` enum in Linux.
        product_version : str, optional
            Installed version of Ansys. The string must be in a format like
            ``"23.2.0"`` (for 2023 R2), matching the documented version format in the
```
Descriptions for parameters use be in noun form and not start with verbs.

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 8]
```diff
@@ -91,6 +92,35 @@ def check_docker_support():
...
+    """Determine the Fluent version during launch_fluent in standalone mode. The version
+    is searched in the following order.
+
+    1. ``product_version`` parameter passed with ``launch_fluent``.
+    2. The latest ANSYS version from ``AWP_ROOTnnn``` environment variables.
+
+    Returns
+    -------
+    FluentVersion, optional
+        Fluent version or None
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:107]:
```suggestion
    """Determine if the Fluent version started by the ''launch_fluent()`` method is in standalone mode.
    
    The version is searched in this order.

    1. The ``product_version`` parameter passed with the ``launch_fluent()`` method.
    2. The latest Ansys version from ``AWP_ROOTnnn``` environment variables.

    Returns
    -------
    FluentVersion, optional
        Fluent version or ``None``.
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 9]
```diff
@@ -91,6 +92,35 @@ def check_docker_support():
...
+    # 1. product_version parameter passed with launch_fluent
+    if product_version:
+        return FluentVersion(product_version)
+
+    # 2. the latest ANSYS version from AWP_ROOT environment variables
+    return FluentVersion.get_latest_installed()
+
+
 def get_fluent_exe_path(**launch_argvals) -> Path:
     """Get Fluent executable path. The path is searched in the following order.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:125]:
```suggestion
    """Get the path for the Fluent executable file.
    
    The search for the path is performed in this order:
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 10]
```diff
@@ -179,6 +209,71 @@ def is_meshing(mode: "FluentMode") -> bool:
         return mode in [FluentMode.MESHING_MODE, FluentMode.PURE_MESHING_MODE]
 
 
+@total_ordering
+class FluentEnum(Enum):
+    """Base class for Fluent-related enums.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:214]:
```suggestion
    """Provides the base class for Fluent-related enums.
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 11]
```diff
@@ -179,6 +209,71 @@ def is_meshing(mode: "FluentMode") -> bool:
...
+            return False
+        for member in type(self):
+            if self == member:
+                return True
+            if other == member:
+                return False
+
+
+class FluentExposure(FluentEnum):
+    """Supported UI or graphics exposure of Fluent."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:247]:
```suggestion
    """Provides supported UI or graphics exposure of Fluent."""
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 12]
```diff
@@ -179,6 +209,71 @@ def is_meshing(mode: "FluentMode") -> bool:
...
+    NULL = ("null",)
+    MSW = ("msw",)
+    DX11 = ("dx11",)
+    OPENGL2 = ("opengl2",)
+    OPENGL = ("opengl",)
+    AUTO = ("",)
+
+
+class FluentLinuxGraphicsDriver(FluentEnum):
+    """Supported graphics driver of Fluent in Linux."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:268]:
```suggestion
    """Provides supported graphics driver of Fluent in Linux."""
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 13]
```diff
@@ -282,28 +383,18 @@ def _get_mode(mode: Optional[Union[FluentMode, str, None]] = None):
...
 
-def _raise_exception_g_gu_in_windows_os(additional_arguments: str) -> None:
-    """If -g or -gu is passed in Windows OS, the exception should be raised."""
-    additional_arg_list = additional_arguments.split()
-    if _is_windows() and (
-        ("-g" in additional_arg_list) or ("-gu" in additional_arg_list)
+def _raise_non_gui_exception_in_windows(
+    exposure: FluentExposure, product_version: FluentVersion
+) -> None:
+    """Exposure < hidden_gui is not supported in Windows in Fluent version < 24.1."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:389]: Exception message should be in plain English with tagged code entities and the less than operator written out.

File: src/ansys/fluent/core/launcher/standalone_launcher.py [hunk 14]
```diff
@@ -64,6 +67,13 @@ def __init__(
         ----------
         mode : FluentMode
             Launch mode of Fluent to point to a specific session type.
+        exposure : FluentExposure
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/standalone_launcher.py:70]: Same here.

File: src/ansys/fluent/core/warnings.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
...
+    """Common warning class for all deprecated PyFluent feature."""
```
Comment (PipKat) [inline src/ansys/fluent/core/warnings.py:2]:
```suggestion
    """Provides the common warning class for all deprecated PyFluent feature."""
```

File: src/ansys/fluent/core/launcher/container_launcher.py (hunk 7 above)
Comment (mkundu1) [reply src/ansys/fluent/core/launcher/container_launcher.py]: Updated, thanks

File: src/ansys/fluent/core/launcher/launcher.py [hunk 16]
```diff
@@ -233,6 +258,27 @@ def launch_fluent(
         raise GPUSolverSupportError()
     _process_kwargs(kwargs)
     del kwargs
+    if show_gui is not None:
+        warnings.warn(
+            "show_gui is deprecated, use exposure instead", PyFluentDeprecationWarning
```
[the remark this answers is not in the document]
Comment (mkundu1) [reply src/ansys/fluent/core/launcher/launcher.py]: Updated, thanks

File: src/ansys/fluent/core/launcher/launcher_utils.py (hunk 8 above)
Comment (mkundu1) [reply src/ansys/fluent/core/launcher/launcher_utils.py]: Here we are determinig the Fluent version from the passed parameters or environment variables, Fluent is not stared at this point. I've reworded the summary line.

File: src/ansys/fluent/core/launcher/launcher_utils.py (hunk 9 above)
Comment (mkundu1) [reply src/ansys/fluent/core/launcher/launcher_utils.py]: updated, thanks

File: src/ansys/fluent/core/launcher/launcher.py [hunk 17]
```diff
@@ -157,6 +167,21 @@ def launch_fluent(
         default is ``True``. You can stop and start the streaming of the
         Fluent transcript subsequently via the method calls, ``transcript.start()``
         and ``transcript.stop()`` on the session object.
+    exposure : FluentExposure or str, optional
+        Exposure of Fluent. Options are either the values of the ``FluentExposure`` enum
+        or any of ``"no_gui_or_graphics"``, ``"no_gui"``, ``"hidden_gui"``,
+        ``"no_graphics"`` or ``"gui"``. The default is ``FluentExposure.HIDDEN_GUI`` in
+        Windows and ``FluentExposure.NO_GUI`` in Linux. ``"no_gui_or_graphics"`` and
+        ``"no_gui"`` exposure are supported in Windows only for Fluent version 2024 R1
+        or later.
```
Comment (raph-luc) [inline src/ansys/fluent/core/launcher/launcher.py:176]: I think people looking for this option will probably be thinking something along the lines of "I want to hide the Fluent interface" or those more familiar "I want to change the interface launch mode of Fluent"

What about calling it `interface` rather than `exposure`? 
```suggestion
    interface: FluentInterface or str, optional
        Interface launch mode of Fluent. Options include the values of the ``FluentInterface`` 
        enum, as well as ``"no_gui_or_graphics"``, ``"no_gui"``, ``"hidden_gui"``,
        ``"no_graphics"`` and ``"gui"``. The default is ``FluentInterface.HIDDEN_GUI`` in
        Windows and ``FluentInterface.NO_GUI`` in Linux. ``"no_gui_or_graphics"`` and
        ``"no_gui"`` interface are supported in Windows only for Fluent version 2024 R1
        or later.
```

Comment (mkundu1) [reply src/ansys/fluent/core/launcher/launcher.py:177]: @raph-luc I feel `interface` can be bit technical for some users. As this is purely a Fluent side option, we can alternatively call it `fluent_ui` so it precisely means what it is for (both `interface` and `exposure` have a broad meaning).

Comment (raph-luc) [reply src/ansys/fluent/core/launcher/launcher.py:177]: @mkundu1 thanks for the explanation, I think that "UI" or "interface" are similarly technical, with "UI" (User Interface) being a bit more technical than "interface". 

Regarding `fluent_ui`, we are already in the `launch_fluent` function, so the `fluent_` prefix is probably unnecessary.

I do personally think that "exposure" is a significantly more technical/obscure term than "interface", hence my suggestion. I guess "window" would be even less technical than "interface", but not sure if it would be appropriate here.

Is there someone from the UX or doc teams that we could briefly ask which one they would suggest?

cc @PipKat @seanpearsonuk

Comment (seanpearsonuk) [reply src/ansys/fluent/core/launcher/launcher.py:177]: I prefer `FluentVisibility` over `FluentDisplayMode`. `FluentExposure` is good although some users could easily think it refers to API exposure (Pro, Enterprise etc), so it would need to be expanded to `FluentInterfaceExposure` (still ambiguous) and therefore `FluentGUIExposure` or the less blatant `FluentUIExposure`. I would be unsure of the meaning of `FluentInterface`.

Comment (raph-luc) [reply src/ansys/fluent/core/launcher/launcher.py:177]: @seanpearsonuk if `FluentInterface` is too broad, what about just `FluentUI` with a respective `ui` argument for `launch_fluent()`, similar to what @mkundu1 proposed earlier? I would argue there no need to add an `Exposure` suffix.

File: src/ansys/fluent/core/launcher/container_launcher.py [hunk 18]
```diff
@@ -59,9 +63,15 @@ def __init__(
         ----------
         mode : FluentMode
             Launch mode of Fluent to point to a specific session type.
+        ui : FluentUI
+            UI option of Fluent. Options are the values of the ``FluentUI`` enum.
```
Comment (raph-luc) [inline src/ansys/fluent/core/launcher/container_launcher.py:67]:
```suggestion
            Fluent user interface mode. Options are the values of the ``FluentUI`` enum.
```

File: src/ansys/fluent/core/launcher/container_launcher.py [hunk 19]
```diff
@@ -59,9 +63,15 @@ def __init__(
...
+        graphics_driver : FluentWindowsGraphicsDriver or FluentLinuxGraphicsDriver
+            Graphics driver of Fluent. Options are the values of the
+            ``FluentWindowsGraphicsDriver`` enum in Windows or the values of the
+            ``FluentLinuxGraphicsDriver`` enum in Linux.
         product_version : str, optional
-            Select an installed version of ANSYS. The string must be in a format like
-            ``"23.2.0"`` (for 2023 R2) matching the documented version format in the
+            Installed version of Ansys. The string must be in a format like
```
Comment (raph-luc) [inline src/ansys/fluent/core/launcher/container_launcher.py:73]:
```suggestion
            Version of Ansys Fluent to launch. The string must be in a format like
```
I think specifying "installed" is a bit redundant. It has to be installed so that it can be launched locally, right? That should be expected by the user.

File: src/ansys/fluent/core/launcher/launcher.py [hunk 20]
```diff
@@ -158,6 +168,21 @@ def launch_fluent(
         default is ``True``. You can stop and start the streaming of the
         Fluent transcript subsequently via the method calls, ``transcript.start()``
         and ``transcript.stop()`` on the session object.
+    ui : FluentUI or str, optional
+        UI option of Fluent. Options are either the values of the ``FluentUI`` enum
```
Comment (raph-luc) [inline src/ansys/fluent/core/launcher/launcher.py:172]:
```suggestion
        Fluent user interface mode. Options are either the values of the ``FluentUI`` enum
```
(same as other suggestion above)

File: src/ansys/fluent/core/launcher/launcher.py [hunk 21]
```diff
@@ -158,6 +168,21 @@ def launch_fluent(
...
+        or any of ``"no_gui_or_graphics"``, ``"no_gui"``, ``"hidden_gui"``,
+        ``"no_graphics"`` or ``"gui"``. The default is ``FluentUI.HIDDEN_GUI`` in
+        Windows and ``FluentUI.NO_GUI`` in Linux. ``"no_gui_or_graphics"`` and
+        ``"no_gui"`` ui options are supported in Windows only for Fluent version 2024 R1
+        or later.
```
Comment (raph-luc) [inline src/ansys/fluent/core/launcher/launcher.py:177]:
```suggestion
        ``"no_gui"`` UI modes are only supported in Windows starting from 
        Fluent version 2024 R1.
```

**Review by raph-luc — APPROVED** [review]
Only minor suggestions for you to consider
